File: C++/Graphs/01_isGraphBipartite.cpp
#include<bits/stdc++.h>
using namespace std;
bool checkBipartiteHelper(int src, vector<vector<int>> &graph, vector<int> color)
{
    queue<int> q;
    q.push(src);
    color[src] = 1;
    while (!q.empty())
    {
        int node = q.front();
        q.pop();

        // traverse all adjacent node
        for (auto it : graph[node])
        {
            if (color[it] == 0)
            {
                color[it] = 3 - color[node];
                q.push(it);
            }
            else if (color[it] == color[node])
            {
                return false;
            }
        }
    }
    return true;
}

bool isBipartite(vector<vector<int>> &graph)
{
    vector<int> color(105, 0);

    for (int i = 0; i < graph.size(); i++)
    {
        if (color[i] == 0)
        {
            if (!checkBipartiteHelper(i, graph, color))
            {
                return false;
            }
        }
    }
    return true;
}

File: Java/Graphs/01_isGraphBipartite.java
public boolean isBipartite(int[][] graph,int src,int[] color){
        LinkedList<Integer> que = new LinkedList<>();
        que.addLast(src);
        int count = 0;
        while(que.size() != 0){
            int size = que.size();
            while(size-->0){
                int vtx = que.removeFirst();
                if(color[vtx] != -1){
                    if(color[vtx] != count)
                        return false;
                    continue;
                }
                color[vtx] = count;
                for(int v: graph[vtx]){
                    if(color[v] == -1){
                        que.addLast(v);
                    }
                }
            }
            count = (count+1)%2;
        }
        return true;    
    }
    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        int color[] = new int[n];
        Arrays.fill(color,-1);
        for(int i=0;i<n;i++){
            if(color[i] == -1){
                if(!isBipartite(graph,i,color)) return false;
            }
        }
        return true;
    }